photon-api/src/main/scala/com/linkedin/photon/ml/data/LocalDataSet.scala

```diff
@@ -212,6 +242,135 @@ object LocalDataSet {
+    */
+
+  protected[ml] def computeRatioCILowerBound(
+      randomLabelAndFeatures: Array[(Double, Vector[Double])],
+      quartile: Double,
+      globalFeatureShardStats: BasicStatisticalSummary): Map[Int, Double] = {
+
+    val dummyBoundForNonBinaryOrIntercept = 2.0
+    val lowerBounds = mutable.Map[Int, Double]()
+    var globalPopulationNumSamples = globalFeatureShardStats.count
+    val globalPopulationFirstOrderMeans = globalFeatureShardStats.mean.toArray
+    val globallFeatureNonZero = globalFeatureShardStats.numNonzeros.toArray
+    val globalMean = globalFeatureShardStats.mean.toArray
+
+    val randomEffectNumSamples: Long = randomLabelAndFeatures.length.toLong
+    val randomFeatureFirstOrderSums = randomLabelAndFeatures.map(_._2).toSeq.reduce(_ + _)
+    
+    val m: Long = randomEffectNumSamples
+    val n: Long = globalPopulationNumSamples
+    val lastColumn = randomFeatureFirstOrderSums.keySet.max
+
+    randomFeatureFirstOrderSums.keySet.foreach { key =>
+      // Do computation on only binary and non-intercept features
+      if (isBinary(globallFeatureNonZero, globalPopulationNumSamples, globalMean, key) && key != lastColumn) {
+        var x: Double = randomFeatureFirstOrderSums(key)
+        var y: Double = globalPopulationFirstOrderMeans(key) * globalPopulationNumSamples
+        var lowerBound = None: Option[Double]
+        var upperBound = None: Option[Double]
+        var py = None: Option[Double]
+        // deal with extreme cases of x and y
+        (x, y) match {
+          case d if (d._1 == 0 && d._2 == 0) => {
+            lowerBound = Some(0.0)
+            upperBound = Some(Double.PositiveInfinity)
+          }
+          case d if (d._1 == 0 && d._2 != 0) => {
+            lowerBound = Some(0.0)
+            x = 0.5
+          }
+          case d if (d._1 != 0 && d._2 == 0) => {
+            y = 0.5
+            py = Some(y / n)
+            upperBound = Some(Double.PositiveInfinity)
+          }
+          case d if (d._1 == m && d._2 == n) => {
+            x = m - 0.5
+            y = n - 0.5
+            py = Some(y / n)
+          }
+          case _ => None
+        }
+        // we have the mean already from the statistics, save computation here
+        py match {
+          case Some(py) => None
+          case _ => py = Some(globalPopulationFirstOrderMeans(key))
+        }
+        val (t,variance) = computeMeanAndVariance(x, y, m, n, py.get)
+
+        upperBound match {
+          case Some(upperBound) => None
+          case _ => {
+            upperBound = Some(computeUpperBound(t, variance, quartile))
+          }
+        }
+
+        lowerBound match {
+          case Some(lowerBound) => None
+          case _ => {
+            lowerBound = Some(computeLowerBound(t, variance, quartile))
+            if (t < 1.0) {
+              lowerBound = Some(1.0 / upperBound.get)
+            }
+          }
+        }
```

Comment:
Actually - we can merge the case where `X = 0` into the first. We can also use the case where `X == m, Y == n` as a shorthand way to keep the intercept around without needing to pass in indices or keep track of it otherwise. I'm thinking something like:
```
if (y == 0 || x == 0) {
  0D
} else if(x == m && y == n) {
  Double.PostiveInfinity
} else {
  val (t, variance) = computeMeanAndVariance(x, m, y, n)
  
  if (t < 1D) {
    1D / computeUpperBound(t, variance, zScore)
  } else {
    computeLowerBound(t, variance, zScore)
  }
}
```
This would also address my earlier comment about the `0.5` constant.